v8go.cc

```diff
@@ -414,6 +414,86 @@ RtnValue RunScript(ContextPtr ctx_ptr, const char* source, const char* origin) {
   return rtn;
 }
 
+static MaybeLocal<Module> ResolveCallback(Local<Context> context,
+                                          Local<String> specifier,
+                                          Local<Module> referrer) {
+  Isolate* isolate = context->GetIsolate();
+  isolate->ThrowException(String::NewFromUtf8(isolate, "import not supported").ToLocalChecked());
+  return MaybeLocal<Module>();
+}
+
+RtnValue RunModule(ContextPtr ctx_ptr, const char* source, const char* origin) {
+  LOCAL_CONTEXT(ctx_ptr);
+
+  Local<String> src =
+      String::NewFromUtf8(iso, source, NewStringType::kNormal).ToLocalChecked();
+  Local<String> ogn =
+      String::NewFromUtf8(iso, origin, NewStringType::kNormal).ToLocalChecked();
+  RtnValue rtn = {nullptr, nullptr};
+
+  ScriptOrigin script_origin(ogn, 0, 0, false, -1, Local<Value>(), false, false, true);
+  ScriptCompiler::Source csrc(src, script_origin);
+  MaybeLocal<Module> mmodule = ScriptCompiler::CompileModule(iso, &csrc);
+  if (mmodule.IsEmpty()) {
+    rtn.error = ExceptionError(try_catch, iso, local_ctx);
+    return rtn;
+  }
+
+  Local<Module> root_module = mmodule.ToLocalChecked();
+  if (!root_module->InstantiateModule(local_ctx, &ResolveCallback).FromMaybe(false)) {
+    rtn.error = ExceptionError(try_catch, iso, local_ctx);
+    return rtn;
+  }
+
+  Local<Value> result;
+  if (!root_module->Evaluate(local_ctx).ToLocal(&result)) {
+    rtn.error = ExceptionError(try_catch, iso, local_ctx);
+    return rtn;
+  }
+
+  // result is a Promise if the module has imports that need to be resolved
+  // OR has a top-level await. imports are not supported yet, but awaits are.
+  //
+  // If await was not used, the result is undefined, and the module's status
+  // should already be kEvaluated.
+  if (result->IsPromise()) {
+    // We have to wait. Just busy-loop and poll for completion.
+    Local<Promise> result_promise(Local<Promise>::Cast(result));
+    while (result_promise->State() == Promise::kPending) {
+      iso->PerformMicrotaskCheckpoint();
+    }
```

Comment:
The top-level await could be blocked on IO done in another go-routine, in which case this spin loop will just waste the thread's CPU time.  The other problem with this loop is that broken promises (those neither resolved or rejected) would result in an infinite loop.

I think the function should just return the promise, which would be consistent with how RunScript doesn't try to wait for promises either.

Reply:
I agree, the result of RunModule should be a Promise. I did some work towards enabling that in https://github.com/rogchap/v8go/pull/119 and https://github.com/rogchap/v8go/pull/118 , but since then my project has adopted an alternative approach to what I was using v8 for, so I'm probably not going to continue this PR. Closing.